File: integrations/android/app/src/test/java/app/airplaystatus/alwayson/AlwaysOnStateTest.kt
package app.airplaystatus.alwayson

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class AlwaysOnStateTest {

    @Test
    fun classifyMapsSnapshots() {
        assertEquals(Mode.PLAYING, classify(isPlaying = true, hasTitle = true))
        assertEquals(Mode.PAUSED, classify(isPlaying = false, hasTitle = true))
        assertEquals(Mode.IDLE, classify(isPlaying = false, hasTitle = false))
    }

    @Test
    fun idleToPlayWakesWithoutNudge() {
        val s = reduce(AlwaysOnState(), AlwaysOnEvent.Playback(isPlaying = true, hasTitle = true))
        assertEquals(Mode.PLAYING, s.mode)
        assertEquals(Screen.AWAKE, s.screen)
        assertFalse(s.nudge)
        assertTrue(shouldKeepScreenOn(s))
    }

    @Test
    fun playThenIdleGraceDimsAndRecordsFocus() {
        var s = reduce(AlwaysOnState(), AlwaysOnEvent.Playback(true, true))
        s = reduce(s, AlwaysOnEvent.Playback(false, false)) // idle
        assertEquals(Screen.AWAKE, s.screen) // grace not elapsed
        s = reduce(s, AlwaysOnEvent.IdleGraceElapsed(focused = true))
        assertEquals(Screen.DIM, s.screen)
        assertTrue(s.focusBeforeIdle)
        assertFalse(shouldKeepScreenOn(s))
    }

    @Test
    fun resumeWhileDimmedAndFocusedShowsNudgeThenDismiss() {
        var s = reduce(AlwaysOnState(), AlwaysOnEvent.IdleGraceElapsed(focused = true))
        s = reduce(s, AlwaysOnEvent.Playback(true, true))
        assertTrue(s.nudge)
        assertEquals(Screen.DIM, s.screen)
        s = reduce(s, AlwaysOnEvent.DismissNudge)
        assertFalse(s.nudge)
        assertEquals(Screen.AWAKE, s.screen)
    }

    @Test
    fun resumeWhileDimmedButNotFocusedIsSilent() {
        var s = reduce(AlwaysOnState(), AlwaysOnEvent.IdleGraceElapsed(focused = false))
        s = reduce(s, AlwaysOnEvent.Playback(true, true))
        assertFalse(s.nudge)
        assertEquals(Screen.AWAKE, s.screen)
    }

    @Test
    fun leavingBeforeDimClearsFocusIntent() {
        var s = reduce(AlwaysOnState(), AlwaysOnEvent.FocusLost)
        assertFalse(s.focusBeforeIdle)
        s = reduce(s, AlwaysOnEvent.IdleGraceElapsed(focused = true))
        s = reduce(s, AlwaysOnEvent.FocusLost) // ignored once dimmed
        assertTrue(s.focusBeforeIdle)
    }
}

File: integrations/deskthing/test/alwaysOnState.test.js
import { test } from 'node:test';
import assert from 'node:assert/strict';
import {
  classifyPlayback,
  initialState,
  reduce,
  shouldKeepScreenOn,
} from '../shared/alwaysOnState.js';

const playing = { isPlaying: true, title: 'Señorita' };
const paused = { isPlaying: false, title: 'Señorita' };
const nothing = { isPlaying: false, title: null };

test('classifyPlayback', () => {
  assert.equal(classifyPlayback(playing), 'playing');
  assert.equal(classifyPlayback(paused), 'paused');
  assert.equal(classifyPlayback(nothing), 'idle');
  assert.equal(classifyPlayback(null), 'idle');
});

test('idle → play (no prior dim) wakes without a nudge', () => {
  let s = initialState();
  s = reduce(s, { type: 'playback', playback: playing });
  assert.equal(s.mode, 'playing');
  assert.equal(s.screen, 'awake');
  assert.equal(s.nudge, false);
  assert.equal(shouldKeepScreenOn(s), true);
});

test('play → idle → grace dims and records focus-before-idle', () => {
  let s = reduce(initialState(), { type: 'playback', playback: playing });
  s = reduce(s, { type: 'playback', playback: nothing });
  assert.equal(s.mode, 'idle');
  assert.equal(s.screen, 'awake'); // grace not elapsed yet
  s = reduce(s, { type: 'idleGraceElapsed', focused: true });
  assert.equal(s.screen, 'dim');
  assert.equal(s.focusBeforeIdle, true);
  assert.equal(shouldKeepScreenOn(s), false);
});

test('resume while dimmed + focus-before-idle → one-shot nudge, dismiss clears it', () => {
  let s = initialState();
  s = reduce(s, { type: 'idleGraceElapsed', focused: true }); // dim (already idle)
  s = reduce(s, { type: 'playback', playback: playing });
  assert.equal(s.nudge, true);
  assert.equal(s.screen, 'dim'); // stays dim until user taps
  s = reduce(s, { type: 'dismissNudge' });
  assert.equal(s.nudge, false);
  assert.equal(s.screen, 'awake');
});

test('resume while dimmed but NOT focus-before-idle → silent wake, no nudge', () => {
  let s = initialState();
  s = reduce(s, { type: 'idleGraceElapsed', focused: false }); // dim, unfocused
  s = reduce(s, { type: 'playback', playback: playing });
  assert.equal(s.nudge, false);
  assert.equal(s.screen, 'awake');
});

test('leaving the app before it dims clears focus-before-idle intent', () => {
  let s = initialState();
  s = reduce(s, { type: 'focusLost' });
  assert.equal(s.focusBeforeIdle, false);
  // after dim, focusLost is ignored (flag already captured)
  s = reduce(s, { type: 'idleGraceElapsed', focused: true });
  s = reduce(s, { type: 'focusLost' });
  assert.equal(s.focusBeforeIdle, true);
});
